### ISSUE TITLE
CI fails for Go 1.15

### ISSUE THREAD

**Issue by marckhouzam** [spf13/cobra#1864]
As was seen in #1863, the go 1.15 CI is failing with 
```
../../../go/pkg/mod/github.com/kyoh86/richgo@v0.3.11/config/load.go:64:9: undefined: os.ReadFile
```
Looks like `richgo` 0.3.11, which was recently released (Nov 15, 2022), requires go 1.16.

**Comment by marckhouzam** [comment]
I see two options here:
1. Fix richgo to v0.3.10 until we drop support for Go 1.15
2. Drop support for Go 1.15 now (as suggested by @umarcor in https://github.com/spf13/cobra/pull/1863#issuecomment-1323567791)

I'm ok with either option.

We also need to update our deprecation document as we have agreed we will not do a major Cobra release when removing support for Go versions.

I suggest our support policy for Go should be something like: "We guarantee to support the latest 2 versions of Go, just like the Go project does.  To give more flexibility to Cobra consumers, we also won't remove support for older Go versions until a valid reason arises; valid reasons are at the discretion of the maintainers and include but are not limited to, hindering the evolution of the project, limiting the use of CI tools, causing a noticeable added burden of maintenance."

What do you think @jpmcb @umarcor ?

**Comment by jpmcb** [comment]
Thanks for opening this discussion Marc - I have a few thoughts:

I think we should remove usage of `richgo` entirely since it only provides us with colored / paged / enriched test output which is non-critical to the cobra library. It also creates a weird barrier to entry to contribute to cobra which I would _personally_ like removed.

If that is unacceptable, I think we should visit deprecating Go 1.15 from cobra. It's been EOL for awhile now so I'm not concerned with breaking consumers of cobra but it does deserve some deep consideration since cobra's posture towards old versions of Go has been "it should just work".

**Comment by marckhouzam** [comment]
> I think we should remove usage of `richgo` entirely since it only provides us with colored / paged / enriched test output which is non-critical to the cobra library. It also creates a weird barrier to entry to contribute to cobra which I would _personally_ like removed.

You had me at “barrier to entry “ 😜

Also, I personally prefer keeping go versions as long as possible, even if they are EOL.  We don't know the reality of the many projects using Cobra so let's help as much as we can.

So +1 to remove our use of richgo.  Thanks for thinking outside the box @jpmcb

**Comment by marckhouzam** [comment]
So @umarcor had the interesting idea of removing richgo for `make test` and therefore removing the barrier to entry, but to add a new `make richtest` target to allow those that like the colored output to continue using it.  I merged the change.

To make the CI logs easy to read, CI was made to point to `make richtest`.  This means CI still fails with Go 1.15 because of richgo.  But @umarcor also posted #1866 to remove Go 1.15 from CI.  Note that #1866 just removes the **build** for Go 1.15, but does not bump the version in `go.mod`, so we still would allow project to use Go 1.15.

I like the approach but would like @jpmcb's opinion.

And thanks @umarcor for your continued efforts.

**Comment by umarcor** [comment]
> We also need to update our deprecation document as we have agreed we will not do a major Cobra release when removing support for Go versions.
>
> I suggest our support policy for Go should be something like: "We guarantee to support the latest 2 versions of Go, just like the Go project does. To give more flexibility to Cobra consumers, we also won't remove support for older Go versions until a valid reason arises; valid reasons are at the discretion of the maintainers and include but are not limited to, hindering the evolution of the project, limiting the use of CI tools, causing a noticeable added burden of maintenance."

See #1807.
Related #1658.

**Comment by thaJeztah** [comment]
Whoops; missed this discussion and I already had a branch created / pushed to update the minimum version to go1.17 (https://github.com/spf13/cobra/pull/1871). Generally, recommendation is indeed to support last 2 versions, although some project may be "slightly" behind, so where possible 3 versions is "nice" to give projects some time-windows to update (but consider it a "best effort").

I noticed the module that's failing specifies go1.17 as minimum version, which matches "3 versions";

```
Error: ../../../go/pkg/mod/github.com/kyoh86/richgo@v0.3.11/config/load.go:64:9: undefined: os.ReadFile
note: module requires Go 1.17
Error: Process completed with exit code 2.
```

**Comment by marckhouzam** [comment]
I went ahead and merged #1866  which stood running CI on go 1.15 (since CI uses richgo which no longer supports go 1.25) but allows projects to continue using Go 1.15.
